First implementation

astralien3000 opened the pull request:
Hello !

I know this repos is not so "alive", but during my work of porting ros2 to microcontroller, I came up with a minimal implementation of rclc. It was compatible with the beta1, which is quite old now, but I tried to update the code.

I'm able to compile it, and to run a publisher and a subscriber. [See the examples.](https://github.com/astralien3000/examples) The publisher in C is able to communicate with a subscriber in C++ for example. But I have some trouble with the subscriber in C, which is crashing when receiving a message.

So, I'm proposing my implementation as a starting point for future work.
Hope It will push forward the C support ^^ !

EDIT : 
Now the subscriber should work properly.

allenh1 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,235 @@
+{
+  rcl_ret_t ret = rcl_init(argc, argv, rcl_get_default_allocator());
+  if(ret != RCL_RET_OK) {
+    puts("ERROR : rcl_init");
```
You might want to use the logger macros instead of a `puts` call here.

allenh1 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,235 @@
+{
+  const char* namespace_ = ""; // TODO : put as parameter
+
+  rclc_node_t* ret = malloc(sizeof(rclc_node_t));
```
Probably should be the `rcl_get_default_allocator()` instead of `malloc`.

mikaelarguedas commented:
Thanks @astralien3000 for contributing on this! We will try to review this soon.

Note that we didn't push the development of the C client library so far because it is still unclear to us how the execution model will work in C and what the final API would look like.

allenh1 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,235 @@
+rclc_ret_t rclc_destroy_node(rclc_node_t * node)
+{
+  rcl_ret_t ret = rcl_node_fini(node->rcl_node);
+  free(node);
```
Symmetric to the above: `rcl_get_default_allocator()->deallocate` instead of `free`.

astralien3000 commented:
I addressed the first comments of @allenh1 :
 - For error messages, I have seen that rclcpp was using fprintf, so I did the same.
 - For allocate, I get a segfault when I use `rcl_get_default_allocator().allocate(size, rcl_get_default_allocator().state)` instead of `malloc`

@mikaelarguedas, I still can make some proposition, even for the API ? ^^
Anyway, I'm also using this implementation to learn and understand better rclcpp and rcl.

mikaelarguedas commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+  if(rc != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_spin_node] error in rcl_wait_set_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
```
I think this should return an error code given that it this failed, any operation on the waitset in the while will crash.

astralien3000 replied:
Yes, and should I destroy the waitset with `rcl_wait_set_fini` in this case ?

mikaelarguedas replied:
In this case, the waitset was not created properly so you shouldn't have to destroy it.

mikaelarguedas commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+    rcl_reset_error();
+  }
+  DEALLOCATE(publisher);
+  return RCL_RET_OK;
```
Should this return `rc` instead of always  OK ?

mikaelarguedas commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+}
+
+rclc_ret_t rclc_publish(const rclc_publisher_t * publisher, const void * ros_message)
+{
```
this should check if `publisher` is `NULL` before trying to access `&publisher->rcl_publisher`

astralien3000 commented on rclc/include/rclc/rclc.h:
```diff
@@ -50,7 +50,7 @@ rclc_destroy_node(rclc_node_t * node);
 rclc_publisher_t *
 rclc_create_publisher(
-  const rclc_node_t * node,
+  rclc_node_t * node,
```
Because the node have an internal list of subscribers, so that it's modified. Well, it's not the case for publishers, but I did also that for publisher for "consistency" reason.
[The same is done with rclcpp](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node.hpp#L127-L133), the node is not const when creating publisher/subscription.

astralien3000 commented on rclc/include/rclc/rclc.h:
```diff
@@ -66,7 +66,7 @@ rclc_publish(const rclc_publisher_t * publisher, const void * ros_message);
+  rclc_node_t * node,
   const rosidl_message_type_support_t * type_support,
   const char * topic_name,
   void (* callback)(const void *),
```
Yes, I did not pay attention to that.

astralien3000 commented on rclc/include/rclc/types.h:
```diff
@@ -21,20 +21,19 @@
+struct rclc_executor_t;
+typedef struct rclc_executor_t rclc_executor_t;
+
+typedef void (* rclc_callback_t)(const void *);
```
Well, it can be a big discussion, but I don't like giving to the user the internal implementation of my structs (the user should only use the functions to manipulate these structs). To my opinion, only the things that are useful for the user should be in the `include` directory.  Also, I usually do that when the memory allocation is done by the library and not by the user. If some definition is useful for several sources files of the library, I would just put a header file in the `src` directory.

astralien3000 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+  return true;
+}
+
+void rclc_sleep_ms(size_t milliseconds)
```
Yes it is used in the [publisher example](https://github.com/astralien3000/examples/blob/master/rclc/minimal_publisher/not_composable.c). But it is tricky. Is changed the implementation compared to beta1, it was supposed to do almost the same things than `rclc_spin_node`, but now it does not. I don't know what rcl_wait do exactly so that I kept it, but I will try to remove it.

I think it is the most "outdated" part of the API, anyway the function should change.

mikaelarguedas replied:
Yeah I think we would need to update the talker example:
https://github.com/ros2/tutorials/blob/master/rclc_tutorials/src/talker.c
to use a timer like the [Python](https://github.com/ros2/demos/blob/master/demo_nodes_py/demo_nodes_py/topics/talker.py) and [c++](https://github.com/ros2/demos/blob/master/demo_nodes_cpp/src/topics/talker.cpp) ones. The difference being that python uses an rcl timer (what I think rclc should use) while rclcpp uses a chrono timer

astralien3000 replied:
Since I have not implemented the timer part, I will go for replacing rclc_sleep_ms by rclc_spin_node_once, enabling user to specify a timeout.

astralien3000 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+void rclc_spin_node(rclc_node_t * node)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc = rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, rcl_get_default_allocator());
```
Oops, yes, my bad, I only tested nodes with 1 subscriber or 1 publisher.

astralien3000 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+      }
+    }
+
+    rc = rcl_wait(&wait_set, RCL_MS_TO_NS(1000));  // 1000ms == 1s, passed as ns
```
I don't know, at least it should not matter for the user.
If the spin function need to handle something else than rcl_wait, the timeout may be lowered, but it's not likely. I'll try to find what rclcpp do.

mikaelarguedas replied:
Looks like rclcpp and rclpy default tto wait forever rather than an arbitrary timeout.
https://github.com/ros2/rclpy/blob/ec2220239d089f7b5d03a6ed1f6a3136e7c4d049/rclpy/rclpy/executors.py#L403
https://github.com/ros2/rclcpp/blob/989084b3de84000a15fb4bd5a7d63929001cd769/rclcpp/include/rclcpp/executor.hpp#L339

rclcpp has a spin function that doesnt expose timeouts to the user and spin_once/spin_some that allow to specify it

rclpy has only a spin_once that default to wait forever and take a use defined timeout

astralien3000 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i]) {
+        char msg[64];
```
Oops, quick and dirty implementation, sorry.
I don't even know how to allocate memory properly for passing the ROS message.
I understand that I [should use the typesupport](https://github.com/ros2/rcl/blob/master/rcl/include/rcl/subscription.h#L199-L255), but still obscure. I'll check that better.
But for the moment, the idea was to pass an over-sized memory block for the ros_message.

astralien3000 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+{
+  (void)queue_size;
+  
+  rclc_publisher_t* ret = ALLOCATE(sizeof(rclc_publisher_t));
```
Yes it would be better, bad habit

astralien3000 commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,255 @@
+          rcl_reset_error();
+        }
+        else {
+          for (size_t i = 0; i < node->subs_s; ++i) {
```
Actually, using the `i` variable is a bug here.
`wait_set.subscriptions[i]` should be constant, and have already been tested non null.
I'm iterating on node->subs, which is non null by construction (for the moment, I need to be careful about this...)

astralien3000 commented:
Hello,

I think I have addressed all comments. `ament test --only rclc` seems happy. Just one question about the copyright line at the beginning of files I have created, do I give the copyright to the `Open Source Robotics Foundation, Inc.` ? (I did that by default)

Also, It seems that using rcl allocators does not crash my code anymore (having a cleaner code helps ! ^^). I can remove the malloc-based implementation. But I think I will keep [the defines](https://github.com/ros2/rclc/pull/8/files#diff-2ac3bdc884a04dc4cdc95db51221e264R24), which make the code shorter.

There is also still [the question about putting rclc structs definitions in `include/` or `src/`](https://github.com/ros2/rclc/pull/8/files#diff-d67cb59d558b3a3f12d6c03e1c71256bR26) that either was not finished or agreed.

allenh1 commented:
> There is also still the question about putting rclc structs definitions in include/ or src/ that either was not finished or agreed.

I'd say put them in `include/`.

> Well, it can be a big discussion, but I don't like giving to the user the internal implementation of my structs (the user should only use the functions to manipulate these structs).

IMHO, I see no need to _hide_ the `struct` from the user; if they are modifying a `struct` directly, they are likely not doing something right, but I don't think it should be explicitly disabled. But, moreover, I can't find anywhere else in the code base where a `struct` or `class` exists outside of a header file (excluding tests).

astralien3000 commented:
I can put it in a header file in `src/`. [rcl has some headers in src](https://github.com/ros2/rcl/blob/ebf6ea1a83d6da291a463f29cdd6e9f5bcfd5229/rcl/src/rcl/common.h). Well it's not for defining structures, but it could be ?

In the current API (`rclc_*_create`), functions handle all the memory allocation. So that there is fewer reasons to show the structs than to hide (hiding prevents users to do weird things). Putting the structs in the include dir is like having a C++ class where all attributes are public.

The only reason to put structs in the include dir is to let the user handle the memory allocation (so that the user is able to use the stack instead of malloc). But to achieve that, we need to modify a little the API : 
```C
// instead of doing that (allocation made by the library)
rclc_node_t * node = rclc_create_node("name", "namespace");

// do that
rclc_node_t node; // allocate on the stack
rclc_init_node(&node, "name", "namespace"); // constructor
```

Actually, the second option looks more like rcl and even rclcpp (but rclcpp allocates on the stack and call the constructor on the same line). For me, both are interesting, it's just a matter of API choice. I could even say that if I want to use this library on microcontroller, I may prefer the second option because it limits the usage of malloc. But since it's using malloc in rmw, that would make very few difference.

I think I have summed up all my arguments ^^. For the moment and with the current API, I would go as I said in my first sentence : structs in a header file, but in `src/`. If you want to modify the API, mabe I would be in an other PR ? Now I let you make the decision (I'll try to not challenge it this time 😄 ).

tfoote commented:
> Just one question about the copyright line at the beginning of files I have created, do I give the copyright to the Open Source Robotics Foundation, Inc. ? (I did that by default)

You're welcome to assign the copyright to OSRF. It generally makes life simpler down the line, so we prefer that if you're willing to do so. Some background here: http://producingoss.com/en/copyright-assignment.html

astralien3000 commented:
Hello,

I finally could not make ament test happy with my solution. Now structs are defined in `types.h`.
Everything is OK from my side, can we consider merging this PR ?
Tell me if there is any other requirement.

mikaelarguedas commented:
> Everything is OK from my side, can we consider merging this PR ?
Tell me if there is any requirement.

Good to know, thanks for iterating, we'll try to review this soon.

We usually don't merge features that don't have unit tests associated with them.
Can you try to send various types of messages back and forth to confirm that this works for any type of message? We have an example of pure rcl communication [here](https://github.com/ros2/system_tests/blob/a4e81e348f7d2b25076f2b11dfde0f073e3850de/test_communication/test/test_messages_c.cpp) that you may be able to adapt to use `rclc` with a some find-replace

mikaelarguedas commented on rclc/CMakeLists.txt:
```diff
@@ -4,8 +4,28 @@ project(rclc)
+set(rclc_sources
+  "src/functions.c"
+)
+set_source_files_properties(${rclc_sources} PROPERTIES LANGUAGE "C")
```
Is this necessary? (I would expect CMake to figure out that `.c` files are c language)
If it's necessary, could we define the CMake project as a C project rather than explicitly setting the source files properties (e.g. https://github.com/ros2/rclpy/blob/74b5f1801e497bad17ab81713bba645e343aeff4/rclpy/CMakeLists.txt#L3)?

mikaelarguedas commented on rclc/CMakeLists.txt:
```diff
@@ -4,8 +4,28 @@ project(rclc)
+  "src/functions.c"
+)
+set_source_files_properties(${rclc_sources} PROPERTIES LANGUAGE "C")
+add_library(${PROJECT_NAME} ${rclc_sources})
```
ROS 2 expects SHARED libraries by default.
To respect the CMake option "BUILD_SHARED_LIBS" we encourage packages to depend on `ament_cmake_ros` to be compiled statically or dynamically according to the CMake option. See https://github.com/ros2/rcl/pull/186/files for an example of converting a package into an `ament_cmake_ros` package

mikaelarguedas commented on rclc/CMakeLists.txt:
```diff
@@ -4,8 +4,28 @@ project(rclc)
 
```
Please add a default C standard for the project like it's done in other packages, see https://github.com/ros2/rcl/blob/293aee7843131dcd4e85d606f6bd4d320d83255b/rcl/CMakeLists.txt#L15-L18 for an example

mikaelarguedas commented on rclc/include/rclc/types.h:
```diff
@@ -15,26 +15,46 @@
+struct rclc_publisher_t
 {
-  // TODO(wjwwood): Fill this out
-} rclc_executor_t;
```
can you keep the references to the executor ? (as this is a structure we will need to implement in the future)

astralien3000 replied:
The exectuor struct is still declared [here](https://github.com/ros2/rclc/pull/8/files/b8d58ad2a208a7831407c384d6a983d4d8a6266e#diff-d67cb59d558b3a3f12d6c03e1c71256bR30). (not defined, though)

mikaelarguedas commented on rclc/include/rclc/type_support.h:
```diff
@@ -0,0 +1,35 @@
+#include "rosidl_generator_c/message_type_support_struct.h"
+
+typedef struct rclc_type_support_t rclc_type_support_t;
+struct rclc_type_support_t
```
maybe `rclc_message_type_support_t` would be more explicit as (I guess) we will need to create a service equivalent in the future.

mikaelarguedas commented on rclc/include/rclc/rclc.h:
```diff
@@ -29,19 +26,19 @@ rclc_init(int argc, char ** argv);
-rclc_ok();
+rclc_ok(void);
 
 /// Sleeps (blocks) for a given number of milliseconds.
```
this docblock should be updated to describe function behavior

mikaelarguedas commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,347 @@
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc =
+    rcl_wait_set_init(&wait_set, node->subs_s, 0, 0, 0, 0, rcl_get_default_allocator());
```
Maybe we should add the [dummy hack](https://github.com/ros2/rclc/pull/8/files#diff-2ac3bdc884a04dc4cdc95db51221e264R70) here as well as the passed node may not have any subscription

mikaelarguedas commented on rclc/src/functions.c:
```diff
@@ -0,0 +1,347 @@
+  const char * topic_name,
+  size_t queue_size)
+{
+  (void)queue_size;
```
Can you please add a TODO here so that we remember to take this parameter into account in a future iteration

astralien3000 commented:
Hello, sorry for the late reply. I have some higher priority projects currently.
I have addressed your last comments, thank you for reviewing.
Still no unit tests, I will work on that.

astralien3000 commented:
Hey, @mikaelarguedas 
I was not able to add and compile the test you pointed me to, so instead I added some unit tests from `rclcpp`. Tell me if it is not sufficient.

I'm able to compile it, but at runtime, I can't link to RMW : 
```
8: >>> [rcutils|error_handling.c:148] rcutils_set_error_state()
8: This error state is being overwritten:
8: 
8:   'failed to find shared library of rmw implementation, at /Users/loicdauphin/ros2_final_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:127'
8: 
8: with this new error message:
8: 
8:   'failed to find shared library of rmw implementation, at /Users/loicdauphin/ros2_final_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:127'
8: 
8: rcutils_reset_error() should be called after error handling to avoid this.
8: <<<
```
While I am able to run the examples correctly.

dirk-thomas commented:
We will decide on the next step on this in the next weeks meeting when @mikaelarguedas is back.

astralien3000 commented:
Hello, sorry for the delay.
I haven't taken time to fix the problem of unit tests.
But i will try to work on it soon.
By the way, I would be happy to know what were your decisions, if there was any ^^

mikaelarguedas commented:
After discussing this internally, we will take contributions on that repository without including it in our CI or our releases. So it will stay commented out in [ros2.repos](https://github.com/ros2/ros2/blob/master/ros2.repos).
Feel free to submit other PRs to this repo in the future, they will be merged on good faith without us reviewing or testing them. You can advertise this client library on the [wiki](https://github.com/ros2/ros2/wiki/ROS-2-Client-Libraries) with the others client libraries developed by third-party to raise visibility.

Thanks for your contribution!
